pkg/controller/servicebindingrequest/servicebinder.go

```diff
@@ -236,40 +241,102 @@ func (b *serviceBinder) onError(
 	return requeueOnNotFound(err, requeueAfter)
 }
 
+// isApplicationSelectorEmpty returns true if applicationSelector is not declared in
+// the Service Binding Request.
+func isApplicationSelectorEmpty(
+	application v1alpha1.ApplicationSelector,
+) bool {
+	var emptyApplication v1alpha1.ApplicationSelector
+	if application == emptyApplication ||
+		application.ResourceRef == "" &&
+			application.LabelSelector.MatchLabels == nil {
+		return true
+	}
+	return false
+}
+
+// handleApplicationError handles scenarios when:
+// 1. applicationSelector not declared in the Service Binding Request
+// 2. application not found
+func (b *serviceBinder) handleApplicationError(reason string, applicationError error) (reconcile.Result, error) {
+	conditionsv1.SetStatusCondition(&b.sbr.Status.Conditions, conditionsv1.Condition{
+		Type:   CollectionReady,
+		Status: corev1.ConditionTrue,
+	})
+	conditionsv1.SetStatusCondition(&b.sbr.Status.Conditions, conditionsv1.Condition{
+		Type:    InjectionReady,
+		Status:  corev1.ConditionFalse,
+		Reason:  reason,
+		Message: applicationError.Error(),
+	})
+
+	// updating status of request instance
+	sbr, err := b.updateStatusServiceBindingRequest(b.sbr, &b.sbr.Status)
+	if err != nil {
+		return requeueOnConflict(err)
+	}
+
+	// appending finalizer, should be later removed upon resource deletion
+	sbr.SetFinalizers(append(removeStringSlice(b.sbr.GetFinalizers(), finalizer), finalizer))
+	if _, err = b.updateServiceBindingRequest(sbr); err != nil {
+		return noRequeue(err)
+	}
+
+	b.logger.Info(applicationError.Error())
+
+	if errors.Is(applicationError, errApplicationNotFound) {
+		sbr.SetFinalizers(nil)
+		if _, err = b.updateServiceBindingRequest(sbr); err != nil {
+			return noRequeue(err)
+		}
+		return requeueOnNotFound(errApplicationNotFound, requeueAfter)
+	}
+
+	return done()
+}
+
 // bind configures binding between the Service Binding Request and its related objects.
 func (b *serviceBinder) bind() (reconcile.Result, error) {
-	sbrStatus := b.sbr.Status.DeepCopy()
-
 	b.logger.Info("Saving data on intermediary secret...")
 
 	secretObj, err := b.secret.createOrUpdate(b.envVars, b.sbr.AsOwnerReference())
 	if err != nil {
 		b.logger.Error(err, "On saving secret data..")
-		return b.onError(err, b.sbr, sbrStatus, nil)
+		return b.onError(err, b.sbr, &b.sbr.Status, nil)
```

Comment:
It seems that all usages of `b.onError` are using a reference to `b.sbr.Status`, so perhaps that parameter can be removed?

```go
return b.onError(err, b.sbr, nil)
```

Reply:
That can be done. It's a good idea. 
But the intention of raising this PR was not to refactor `b.onError(..)`
We can create an issue for that and tackle it in another PR.